iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,255 @@
+/// </summary>
+/// <param name="telemetryName">The name of the telemetry, as defined in the DTDL interface.</param>
+/// <param name="telemetryValue">The telemetry payload, in the format defined in the DTDL interface.</param>
+/// <param name="componentName">The name of the component in which the telemetry is defined. Can be null for telemetry defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SendTelemetryAsync(string telemetryName, dynamic telemetryValue, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Send a batched instance of telemetry.
+/// </summary>
+/// <param name="telemetryDictionary">The name and value telemetry pairs, as defined in the DTDL interface.</param>
+/// <param name="componentName">The name of the component in which the telemetry is defined. Can be null for telemetry defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SendTelemetryAsync(IDictionary<string, dynamic> telemetryDictionary, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Send a single instance of telemetry, formatted as per DTDL specifications:
+/// - the payload should be in the format: { "<telemetryName>" : "<telemetry_value>" }
+/// - the property Message.ContentEncoding should be set to "utf-8".
+/// - the property Message.ContentType should be set to "application/json".
+/// - if the telemetry is defined under a component, the property Message.ComponentName should be set to "<component_name>".
+/// </summary>
+/// <param name="telemetryMessage">The telemetry message to be sent.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SendTelemetryAsync(Message telemetryMessage, CancellationToken cancellationToken = default);
+```
+
+### Commands
+
+```csharp
+/// <summary>
+/// Subscribe to command event.
+/// </summary>
+/// <param name="commandName">The name of the command, as defined in the DTDL interface.</param>
+/// <param name="commandCallback">The action to be taken when a command request is received.</param>
+/// <param name="userContext">Context object that is passed to the event callback.</param>
+/// <param name="componentName">The name of the component in which the command is defined. Can be null for command defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SubscribeToCommandEvent(string commandName, Func<CommandRequest, object, Task<CommandResponse>> commandCallback, object userContext = default, string componentName = default, CancellationToken cancellationToken = default);
+```
+
+<details>
+<summary>CommandRequest</summary>
+
+```csharp
+public sealed class CommandRequest : MethodRequest
```

Comment:
Maybe really dumb jspaith vs C# coming up -- will this even work?  MethodRequest is sealed which I thought meant it couldn't be inherited?

Reply:
You're right, we cannot inherit sealed classes. 
The idea here is that if we decide to create CommandRequest/Response as a simple wrapper over the existing types (since the differences are minor) we would remove the `sealed` keyword from the existing types.